File: app/src/main/java/network/minter/bipwallet/internal/views/list/multirow/MultiRowAdapter.java
/*
 * Copyright (C) 2018 by MinterTeam
 * @link https://github.com/MinterTeam
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.internal.views.list.multirow;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v4.util.SimpleArrayMap;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.annimon.stream.Stream;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.WeakHashMap;

import timber.log.Timber;

import static network.minter.bipwallet.internal.common.Preconditions.checkNotNull;

/**
 * Mds. 2017
 *
 * @author Eduard Maximovich <edward.vstock@gmail.com>
 */
public class MultiRowAdapter extends RecyclerView.Adapter<MultiRowAdapter.RowViewHolder> {

    //@TODO может weak ссылки хранить?
    protected List<MultiRowContract.Row> mItems = new ArrayList<>();
    private SimpleArrayMap<Integer, Class<? extends RowViewHolder>> holderViewIdClassCache = new SimpleArrayMap<>();
    private WeakHashMap<Integer, MultiRowContract.Row<?>> rowsViewIdClassCache = new WeakHashMap<>();
    private LayoutInflater layoutInflater;
    private boolean mEnableSorting = true;

    public MultiRowAdapter(Builder builder) {
        this(builder.mRows);
        mEnableSorting = builder.mEnableSort;
        if (mEnableSorting) {
            sort();
        }
    }

    public MultiRowAdapter(@NonNull ArrayList<MultiRowContract.Row> items) {
        mItems = checkNotNull(items, "Rows can't be a null");
        makeHoldersCache();
    }

    public MultiRowAdapter() {
    }

    public void updateRows(List<MultiRowContract.Row<?>> newRows) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffCallback(newRows), true);
        diffResult.dispatchUpdatesTo(this);
    }

    public <V extends MultiRowAdapter.RowViewHolder, T extends MultiRowContract.Row<V>>
    MultiRowAdapter addRow(T row, int position) {
        if (row == null || !row.isVisible()) {
            return this;
        }

        int beforeSize = mItems.size();
        final SortableRow<V, T> newRow = new SortableRow<>(row, position);
        if (mItems.contains(newRow)) {
            Timber.d("Updating row: %s with position %d", row.getClass().getSimpleName(), position);
            final int idx = mItems.indexOf(newRow);
            mItems.set(idx, newRow);
            notifyItemChanged(idx);
            return this;
        }
        Timber.d("Adding row: %s with position %d", row.getClass().getSimpleName(), position);

        mItems.add(newRow);

        if (mEnableSorting) {
            sort();
        }

        makeHoldersCache();
        if (beforeSize == 0) {
            notifyItemInserted(0);
        } else {
            notifyItemRangeChanged(0, mItems.size());
        }
        return this;
    }

    @SuppressWarnings("Convert2MethodRef")
    public void addRowsTop(Collection<MultiRowContract.Row> rows) {
        if (rows.isEmpty()) return;

        final List<MultiRowContract.Row> target = Stream.of(rows)
                .filter(item -> item != null)
                .filter(MultiRowContract.Row::isVisible)
                .toList();

        mItems.addAll(0, target);
        if (mEnableSorting) {
            sort();
        }

        makeHoldersCache();
        notifyItemRangeInserted(0, target.size());
    }

    @SuppressWarnings("Convert2MethodRef")
    public void addRows(Collection<MultiRowContract.Row> rows) {
        if (rows.isEmpty()) return;

        final List<MultiRowContract.Row> target = Stream.of(rows)
                .filter(item -> item != null)
                .filter(MultiRowContract.Row::isVisible)
                .toList();

        int beforeSize = mItems.size();
        mItems.addAll(target);
        if (mEnableSorting) {
            sort();
        }

        makeHoldersCache();
        notifyItemRangeInserted(beforeSize, target.size());
    }

    public void addRowTop(MultiRowContract.Row row) {
        if (row == null || !row.isVisible()) {
            return;
        }
        mItems.add(0, row);
        if (mEnableSorting) {
            sort();
        }

        makeHoldersCache();
        notifyItemInserted(mItems.size());
    }

    public void addRow(MultiRowContract.Row row) {
        if (row == null || !row.isVisible()) {
            return;
        }
        mItems.add(row);
        if (mEnableSorting) {
            sort();
        }

        makeHoldersCache();
        notifyItemInserted(mItems.size());
    }

    public void setEnableSorting(boolean enableSorting) {
        mEnableSorting = enableSorting;
    }

    @NonNull
    @Override
    public RowViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final String tag = String.format("TxSendCoinViewHolder instancing: %s", parent.getContext().getResources().getResourceEntryName(viewType));
//        TimeProfiler.start(tag);

        if (layoutInflater == null) {
            layoutInflater = LayoutInflater.from(parent.getContext());
        }

        if (viewType == View.NO_ID || viewType == 0) {
            throw new RuntimeException("Layout id can't be 0");
        }

        View v = layoutInflater.inflate(viewType, parent, false);

        RowViewHolder viewHolder = null;
        Throwable cause = null;
        try {
            viewHolder = findViewHolder(viewType, v);
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException | InstantiationException e) {
            cause = e;
            Timber.e(e, "Error finding view holder");
        }

        if (viewHolder == null) {
            throw new RuntimeException(cause);
        }

        return viewHolder;
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onBindViewHolder(@NonNull RowViewHolder holder, int position) {
        MultiRowContract.Row item = getItemByPosition(position);
        item.onBindViewHolder(holder);
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onViewDetachedFromWindow(@NonNull RowViewHolder holder) {
        super.onViewDetachedFromWindow(holder);
        int position = holder.getAdapterPosition();
        if (position < 0) {
            return;
        }

        MultiRowContract.Row item = getItemByPosition(position);
        try {
            item.onUnbindViewHolder(holder);
        } catch (ClassCastException ignore) {
            // @TODO fix this (throws after removing row)
        }
    }

    @Override
    public int getItemViewType(int position) {
        return getItemByPosition(position).getItemView();
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public void clear() {
        if (mItems.isEmpty()) return;
        mItems.clear();
        notifyDataSetChanged();
        holderViewIdClassCache.clear();
    }

    public void clearNotify() {
        final int size = mItems.size();
        if (size == 0) return;

        notifyItemRangeRemoved(0, size);
        mItems.clear();
    }

    /**
     * Nullptr safe delete row
     *
     * @param row MultiRowContract.Row
     * @param <T> Result
     */
    @SuppressWarnings("unchecked")
    public <T extends MultiRowContract.Row> void remove(T row) {
        if (row == null) return;
        if (mItems.contains(row)) {
            int index = mItems.indexOf(row);
            mItems.remove(index);
            notifyItemRemoved(index);
        } else {
            T found = null;
            for (MultiRowContract.Row r : mItems) {
                if (r instanceof SortableRow && ((SortableRow) r).getRow().equals(row)) {
                    found = (T) r;
                    break;
                }
            }

            remove(found);
        }
    }

    public void removeUniqueLayout(@LayoutRes int rowLayout) {
        if (rowLayout == View.NO_ID) return;

        for (MultiRowContract.Row r : mItems) {
            if (r.getItemView() == rowLayout) {
                remove(r);
                return;
            }
        }
    }

    public <T extends MultiRowContract.Row> void removeUniqueLayout(T row) {
        if (row == null) return;

        for (MultiRowContract.Row r : mItems) {
            if (r.getItemView() == row.getItemView()) {
                remove(r);
                return;
            }
        }
    }

    public MultiRowContract.Row getItemByPosition(int position) {
        if (position < 0 || position >= mItems.size()) {
            return null;
        }
        return mItems.get(position);
    }

    public <T> Stream<T> findStream(Class<T> rowClass) {
        return Stream.of(mItems)
                .filter(item -> {
                    if (item instanceof SortableRow) {
                        return rowClass.isInstance(item) || rowClass.isInstance(((SortableRow) item).getRow());
                    }

                    return rowClass.isInstance(item);
                })
                .map(rowClass::cast);
    }

    public void sort(Comparator<MultiRowContract.Row> c) {
        Collections.sort(mItems, c);
    }

    public void sort() {
        Collections.sort(mItems, new RowComparator());
    }


    /**
     * Кэшированный список классов TxSendCoinViewHolder'ов чтоб итеративно каждый раз не искать
     */
    @SuppressWarnings("unchecked")
    protected void makeHoldersCache() {
        checkNotNull(mItems, "Wow! Rows can't be null");
        holderViewIdClassCache.clear();
        holderViewIdClassCache = new SimpleArrayMap<>(mItems.size());
        rowsViewIdClassCache = new WeakHashMap<>(mItems.size());

        for (MultiRowContract.Row item : mItems) {
            checkNotNull(item);
            if (item instanceof SortableRow) {
                checkNotNull(item.getViewHolderClass(),
                        "Row " + (((SortableRow) item).getRow().getClass()) + " does not have valid TxSendCoinViewHolder class");
            } else {
                checkNotNull(item.getViewHolderClass(),
                        "Row " + item.getClass() + " does not have valid TxSendCoinViewHolder class");
            }

            rowsViewIdClassCache.put(item.getItemView(), item);
            holderViewIdClassCache.put(item.getItemView(), item.getViewHolderClass());
        }
    }

    private boolean isInnerClass(Class<?> clazz) {
        return clazz.isMemberClass() && !Modifier.isStatic(clazz.getModifiers());
    }

    private MultiRowContract.Row<?> findRow(@LayoutRes int viewId) {
        final MultiRowContract.Row<?> row = rowsViewIdClassCache.get(viewId);
        if (row == null) {
            makeHoldersCache();
        }

        return rowsViewIdClassCache.get(viewId);
    }

    @NonNull
    private RowViewHolder findViewHolder(@LayoutRes int viewId, View view)
            throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
        Class<? extends RowViewHolder> holderClass = holderViewIdClassCache.get(viewId);
        if (holderClass == null) {
            makeHoldersCache();
        }
        holderClass = holderViewIdClassCache.get(viewId);
        if (holderClass == null) {
            throw new RuntimeException("Can't findStream TxSendCoinViewHolder for view " + String.valueOf(viewId));
        }
        if (isInnerClass(holderClass)) {
            throw new RuntimeException("Class should be static!");
        }
        return holderClass.getDeclaredConstructor(View.class).newInstance(view);
    }

    public static class Builder {

        private ArrayList<MultiRowContract.Row> mRows = new ArrayList<>();
        private boolean mEnableSort;

        public Builder addRow(MultiRowContract.Row row) {
            mRows.add(row);
            return this;
        }

        public Builder enableSort(boolean enable) {
            mEnableSort = enable;
            return this;
        }

        public MultiRowAdapter build() {
            return new MultiRowAdapter(this);
        }
    }

    public static abstract class RowViewHolder extends RecyclerView.ViewHolder {

        public RowViewHolder(View itemView) {
            super(itemView);
        }
    }

    public static class RowComparator implements Comparator<MultiRowContract.Row> {

        @Override
        public int compare(MultiRowContract.Row o1, MultiRowContract.Row o2) {
            return o1.getRowPosition() - o2.getRowPosition();
        }
    }

    final class DiffCallback extends DiffUtil.Callback {
        List<MultiRowContract.Row<?>> mRows;

        DiffCallback(List<MultiRowContract.Row<?>> rows) {
            mRows = rows;
        }

        @Override
        public int getOldListSize() {
            return mItems.size();
        }

        @Override
        public int getNewListSize() {
            return mRows.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return mItems.get(oldItemPosition).getItemView() == mRows.get(newItemPosition).getItemView()
                    && mItems.get(oldItemPosition).getRowPosition() == mRows.get(newItemPosition).getRowPosition();
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            return mItems.get(oldItemPosition).equals(mRows.get(newItemPosition));
        }
    }
}

File: app/src/main/java/network/minter/bipwallet/internal/views/list/multirow/MultiRowAdapter.kt
/*
 * Copyright (C) by MinterTeam. 2020
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package network.minter.bipwallet.internal.views.list.multirow

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.collection.SimpleArrayMap
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.annimon.stream.Stream
import network.minter.bipwallet.internal.common.Preconditions
import network.minter.bipwallet.internal.views.list.multirow.MultiRowAdapter.RowViewHolder
import network.minter.bipwallet.internal.views.list.multirow.MultiRowContract.Row
import timber.log.Timber
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Modifier
import java.util.*

/**
 * Mds. 2017
 *
 * @author Eduard Maximovich <edward.vstock></edward.vstock>@gmail.com>
 */
open class MultiRowAdapter : RecyclerView.Adapter<RowViewHolder> {

    protected var mItems: MutableList<Row<RowViewHolder>> = ArrayList()

    private var holderViewIdClassCache = SimpleArrayMap<Int, Class<out RowViewHolder>>()
    private var rowsViewIdClassCache = WeakHashMap<Int, Row<RowViewHolder>>()
    private var layoutInflater: LayoutInflater? = null
    private var mEnableSorting = true

    constructor(builder: Builder) : this(builder.mRows) {
        mEnableSorting = builder.mEnableSort
        if (mEnableSorting) {
            sort()
        }
    }


    @Suppress("UNCHECKED_CAST")
    constructor (items: MutableList<Row<*>>) {
        mItems = items as MutableList<Row<RowViewHolder>>
        makeHoldersCache()
    }

    constructor()

    fun updateRows(newRows: List<Row<*>>) {
        val diffResult = DiffUtil.calculateDiff(DiffCallback(newRows), true)
        diffResult.dispatchUpdatesTo(this)
    }


    @Suppress("UNCHECKED_CAST")
    fun <R : RowViewHolder, T : Row<R>> addRow(row: T, position: Int): MultiRowAdapter {
        if (!row.isVisible()) {
            return this
        }
        val beforeSize = mItems.size
        val newRow = SortableRow(row, position)
        if (mItems.contains(newRow as Row<*>)) {
            Timber.d("Updating row: %s with position %d", row.javaClass.simpleName, position)
            val idx = mItems.indexOf(newRow as Row<*>)
            mItems[idx] = newRow as Row<RowViewHolder>
            notifyItemChanged(idx)
            return this
        }
        Timber.d("Adding row: %s with position %d", row.javaClass.simpleName, position)
        mItems.add(newRow as Row<RowViewHolder>)
        if (mEnableSorting) {
            sort()
        }
        makeHoldersCache()
        if (beforeSize == 0) {
            notifyItemInserted(0)
        } else {
            notifyItemRangeChanged(0, mItems.size)
        }
        return this
    }

    @Suppress("UNCHECKED_CAST")
    fun addRowsTop(rows: Collection<Row<*>>) {
        if (rows.isEmpty()) return

        val targets = rows.filter {
            it.isVisible()
        }.map {
            it as Row<RowViewHolder>
        }

        mItems.addAll(0, targets)
        if (mEnableSorting) {
            sort()
        }
        makeHoldersCache()
        notifyItemRangeInserted(0, targets.size)
    }

    @Suppress("UNCHECKED_CAST")
    fun addRows(rows: Collection<Row<*>>) {
        if (rows.isEmpty()) return

        val targets = rows.filter { it.isVisible() }.map { it as Row<RowViewHolder> }

        val beforeSize = mItems.size
        mItems.addAll(targets)
        if (mEnableSorting) {
            sort()
        }
        makeHoldersCache()
        notifyItemRangeInserted(beforeSize, targets.size)
    }

    @Suppress("UNCHECKED_CAST")
    fun addRowTop(row: Row<*>?) {
        if (row == null || !row.isVisible()) {
            return
        }
        mItems.add(0, row as Row<RowViewHolder>)
        if (mEnableSorting) {
            sort()
        }
        makeHoldersCache()
        notifyItemInserted(mItems.size)
    }

    @Suppress("UNCHECKED_CAST")
    fun addRow(row: Row<*>) {
        if (!row.isVisible()) {
            return
        }
        mItems.add(row as Row<RowViewHolder>)
        if (mEnableSorting) {
            sort()
        }
        makeHoldersCache()
        notifyItemInserted(mItems.size)
    }

    fun setEnableSorting(enableSorting: Boolean) {
        mEnableSorting = enableSorting
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RowViewHolder {
//        val tag = String.format("RowViewHolder instancing: %s", parent.context.resources.getResourceEntryName(viewType))
        //        TimeProfiler.start(tag);
        if (layoutInflater == null) {
            layoutInflater = LayoutInflater.from(parent.context)
        }
        if (viewType == View.NO_ID || viewType == 0) {
            throw RuntimeException("Layout id can't be 0")
        }
        val v = layoutInflater!!.inflate(viewType, parent, false)
        var viewHolder: RowViewHolder? = null
        var cause: Throwable? = null
        try {
            viewHolder = findViewHolder(viewType, v)
        } catch (e: NoSuchMethodException) {
            cause = e
            Timber.e(e, "Error finding view holder")
        } catch (e: IllegalAccessException) {
            cause = e
            Timber.e(e, "Error finding view holder")
        } catch (e: InvocationTargetException) {
            cause = e
            Timber.e(e, "Error finding view holder")
        } catch (e: InstantiationException) {
            cause = e
            Timber.e(e, "Error finding view holder")
        }
        if (viewHolder == null) {
            throw RuntimeException(cause)
        }
        return viewHolder
    }

    override fun onBindViewHolder(holder: RowViewHolder, position: Int) {
        val item: Row<RowViewHolder> = getItemByPosition(position)
        item.onBindViewHolder(holder)
    }

    override fun onViewDetachedFromWindow(holder: RowViewHolder) {
        super.onViewDetachedFromWindow(holder)
        val position = holder.bindingAdapterPosition
        if (position < 0) {
            return
        }
        val item = getItemByPosition(position)
        try {
            item.onUnbindViewHolder(holder)
        } catch (ignore: ClassCastException) {
            // @TODO fix this (throws after removing row)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return getItemByPosition(position).getItemView()
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    fun clear() {
        if (mItems.isEmpty()) return
        mItems.clear()
        notifyDataSetChanged()
        holderViewIdClassCache.clear()
    }

    fun clearNotify() {
        val size = mItems.size
        if (size == 0) return
        notifyItemRangeRemoved(0, size)
        mItems.clear()
    }

    /**
     * Nullptr safe delete row
     *
     * @param row MultiRowContract.Row
     * @param <T> Result
    </T> */
    fun remove(row: Row<*>) {
        if (mItems.contains(row)) {
            val index = mItems.indexOf(row)
            mItems.removeAt(index)
            notifyItemRemoved(index)
        } else {
            var found: Row<*>? = null
            for (r in mItems) {
                if (r is SortableRow<*, *> && r.row == row) {
                    found = r
                    break
                }
            }
            if (found != null) {
                remove(found)
            }

        }
    }

    fun removeUniqueLayout(@LayoutRes rowLayout: Int) {
        if (rowLayout == View.NO_ID) return
        for (r in mItems) {
            if (r.getItemView() == rowLayout) {
                remove(r)
                return
            }
        }
    }

    fun removeUniqueLayout(row: Row<*>) {
        for (r in mItems) {
            if (r.getItemView() == row.getItemView()) {
                remove(r)
                return
            }
        }
    }

    fun getItemByPosition(position: Int): Row<RowViewHolder> {
        return if (position < 0 || position >= mItems.size) {
            throw IllegalAccessException("Invalid position $position")
        } else mItems[position]
    }

    fun <T> findStream(rowClass: Class<T>): Stream<T> {
        return Stream.of(mItems)
                .filter { item: Row<*>? ->
                    if (item is SortableRow<*, *>) {
                        return@filter rowClass.isInstance(item) || rowClass.isInstance(item.row)
                    }
                    rowClass.isInstance(item)
                }
                .map { obj: Row<*>? -> rowClass.cast(obj) }
    }

    fun sort(c: Comparator<Row<*>?>?) {
        Collections.sort(mItems, c)
    }

    fun sort() {
        Collections.sort(mItems, RowComparator())
    }

    /**
     * Кэшированный список классов RowViewHolder чтоб итеративно каждый раз не искать
     */
    fun makeHoldersCache() {
        Preconditions.checkNotNull<List<Row<*>>>(mItems, "Wow! Rows can't be null")
        holderViewIdClassCache.clear()
        holderViewIdClassCache = SimpleArrayMap(mItems.size)
        rowsViewIdClassCache = WeakHashMap(mItems.size)
        for (item in mItems) {
            Preconditions.checkNotNull(item)
            if (item is SortableRow<*, *>) {
                Preconditions.checkNotNull<Class<*>>(item.getViewHolderClass(),
                        "Row " + item.row.javaClass + " does not have valid RowViewHolder class")
            } else {
                Preconditions.checkNotNull<Class<*>>(item.getViewHolderClass(),
                        "Row " + item.javaClass + " does not have valid RowViewHolder class")
            }
            rowsViewIdClassCache[item.getItemView()] = item
            holderViewIdClassCache.put(item.getItemView(), item.getViewHolderClass())
        }
    }

    private fun isInnerClass(clazz: Class<*>): Boolean {
        return clazz.isMemberClass && !Modifier.isStatic(clazz.modifiers)
    }

    private fun findRow(@LayoutRes viewId: Int): Row<*> {
        val row = rowsViewIdClassCache[viewId]
        if (row == null) {
            makeHoldersCache()
        }
        return rowsViewIdClassCache[viewId]!!
    }

    @Throws(NoSuchMethodException::class, IllegalAccessException::class, InvocationTargetException::class, InstantiationException::class)
    private fun findViewHolder(@LayoutRes viewId: Int, view: View): RowViewHolder {
        var holderClass = holderViewIdClassCache[viewId]
        if (holderClass == null) {
            makeHoldersCache()
        }
        holderClass = holderViewIdClassCache[viewId]
        if (holderClass == null) {
            throw RuntimeException("Can't findStream RowViewHolder for view $viewId")
        }
        if (isInnerClass(holderClass)) {
            throw RuntimeException("Class should be static!")
        }
        return holderClass.getDeclaredConstructor(View::class.java).newInstance(view)
    }

    class Builder {
        val mRows = ArrayList<Row<*>>()
        var mEnableSort = false
        fun addRow(row: Row<*>): Builder {
            mRows.add(row)
            return this
        }

        fun enableSort(enable: Boolean): Builder {
            mEnableSort = enable
            return this
        }

        fun build(): MultiRowAdapter {
            return MultiRowAdapter(this)
        }
    }

    abstract class RowViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    class RowComparator : Comparator<Row<*>> {
        override fun compare(o1: Row<*>, o2: Row<*>): Int {
            return o1.getRowPosition() - o2.getRowPosition()
        }
    }

    internal inner class DiffCallback(var mRows: List<Row<*>>) : DiffUtil.Callback() {
        override fun getOldListSize(): Int {
            return mItems.size
        }

        override fun getNewListSize(): Int {
            return mRows.size
        }

        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return (mItems[oldItemPosition].getItemView() == mRows[newItemPosition].getItemView()
                    && mItems[oldItemPosition].getRowPosition() == mRows[newItemPosition].getRowPosition())
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return mItems[oldItemPosition] == mRows[newItemPosition]
        }

    }
}